adcc/LazyMp.py

```diff
@@ -60,39 +57,80 @@ def __getattr__(self, attr):
 
     def df(self, space):
         """Delta Fock matrix"""
-        # TODO This is so cheap to compute that it does not really deserve
-        #      any caching
-        return self.__cpp.df(space)
+        hf = self.reference_state
+        s1 = space[:2]
+        s2 = space[2:]
+        fC = hf.fock(s1 + s1).diagonal()
+        fv = hf.fock(s2 + s2).diagonal()
+        return direct_sum("-i+a->ia", fC, fv).evaluate()
 
+    @cached_member_function
     def t2(self, space):
         """T2 amplitudes"""
-        return self.__cpp.t2(space)
-
-    def set_t2(self, space, tensor):
-        """
-        Set the T2 amplitudes tensor. This invalidates all data depending on the T2
-        amplitudes in this class. Note, that potential other caches, such as
-        computed ADC intermediates are *not* automatically invalidated.
-        """
-        return self.__cpp.set_t2(space, tensor)
+        hf = self.reference_state
+        if space == b.oovv:
+            eia = self.df(b.ov)
+            return compute_t2(hf.oovv, eia, eia)
+        elif space == b.ocvv:
+            eia = self.df(b.ov)
+            eca = self.df(b.cv)
+            return compute_t2(hf.ocvv, eia, eca)
+        elif space == b.ccvv:
+            eca = self.df(b.cv)
+            return compute_t2(hf.ccvv, eca, eca)
+        else:
+            raise NotImplementedError("T2 amplitudes not "
+                                      f"implemented for space {space}.")
 
+    @cached_member_function
     def td2(self, space):
         """Return the T^D_2 term"""
-        return self.__cpp.td2(space)
+        if space != b.oovv:
+            raise NotImplementedError("T^D_2 term not implemented "
+                                      f"for space {space}.")
+        t2erit = self.t2eri(b.oovv, b.ov).transpose((1, 0, 2, 3))
+        denom = direct_sum(
+            'ia,jb->ijab', self.df(b.ov), self.df(b.ov)
+        ).symmetrise(0, 1)
+        return evaluate(
+            (+ 4.0 * t2erit.antisymmetrise(2, 3).antisymmetrise(0, 1)
+             - 0.5 * self.t2eri(b.oovv, b.vv)
+             - 0.5 * self.t2eri(b.oovv,  b.oo)) / denom
```

Comment:
Just to check: `/ denom` does not screw up the symmetry, i.e. libadc does the same thing here?

Reply:
Also two spaces before `b.oo`

Reply:
libadc uses `div`, so I guess we're good.